File: BAEKJOON/src/silver/iii/java/BOJ_2606_Virus.java
package silver.iii.java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

// 바이러스
public class BOJ_2606_Virus {

    static void dfs(List<List<Integer>> graph, int root, boolean[] visited) {

        visited[root] = true;
        for (int i = 0; i < graph.get(root).size(); i++) {
            int next = graph.get(root).get(i);
            if (!visited[next]) {
                dfs(graph, next, visited);
            }
        }
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        List<List<Integer>> graph = new ArrayList<>(n);
        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            String[] input = br.readLine().split(" ");
            int u = Integer.parseInt(input[0]);
            int v = Integer.parseInt(input[1]);
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        boolean[] visited = new boolean[n + 1];
        dfs(graph, 1, visited);
        int count = 0;
        for (int i = 2; i < n + 1; i++) {
            if (visited[i]) {
                count++;
            }
        }
        System.out.println(count);
    }
}

File: BAEKJOON/src/silver/iii/py/BOJ_2606_Virus.py
# 바이러스
import sys
# from collections import deque


# def bfs(graph, root, visited):
#     queue = deque([root])
#     visited[root] = True
#     while queue:
#         v = queue.popleft()
#         for i in graph[v]:
#             if not visited[i]:
#                 queue.append(i)
#                 visited[i] = True
#     return visited.count(True) - 1


def dfs(graph, root, visited=None):
    if visited is None:
        visited = []
    visited.append(root)
    for node in graph[root]:
        if node not in visited:
            dfs(graph, node, visited)
    return len(visited) - 1


n = int(sys.stdin.readline())
m = int(sys.stdin.readline())
graph = [[] for _ in range(n + 1)]
# visited = [False] * (n + 1)
for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)
# print(bfs(graph, 1, visited))
print(dfs(graph, 1))
